### ISSUE TITLE
sample_weight

### ISSUE THREAD

**Issue by darkvoyager** [dnouri/skorch#203]
Hi guys, it would be useful to support 'sample_weight' which is supported by many sklearn models. In order to do that, you would need to modify the loss fonction (pass reduce=False to the pytorch loss function -> pytorch function returns the loss for each individual samples), then take the dot product of the loss per sample times the sample weight. That also means, that when splitting the data between training & validation and for the mini-batch, the split needs to be aware of the sample_weight (not just X and y). So that's a little bit of work, but nothing major I think.

**Comment by BenjaminBossan** [comment]
Thanks for the suggestion. In addition to the points you mentioned, we would also have to make sure that the sample weights are batched, so it is no trivial task. We will look at it.

For the time being, the easiest way to get the same functionality would be to pass the sample weights as part of your `X` and override `get_loss` to do the weighting:

```
X, y = get_data()
X = {'X': X, 'sample_weight': sample_weight}

class MyModule(nn.Module):
    ...
    def forward(self, X, sample_weight):
        # sample_weight is passed to forward
        ...

class MyNet(NeuralNet):
    def get_loss(self, y_pred, y_true, X, *args, **kwargs):
        loss_unreduced = super().get_loss(y_pred, y_true, X, *args, **kwargs)
        sample_weight = X['sample_weight']
        return (sample_weight * loss_unreduced).mean()

net = MyNet(MyModule, ..., criterion__reduce=False)
net.fit(X, y)
```

**Comment by benjamin-work** [comment]
Maybe you could try my version above, since it should take care of train/valid splitting and batching. For your forward step, since you usually don't need the sample weights there, you could just bury it in kwargs: `def forward(self, X, **kwargs):`.

**Comment by darkvoyager** [comment]
Got it, your solution is much better indeed, thanks a lot.

**Comment by benjamin-work** [comment]
I added a section to the FAQ (#221) that is a cleaned-up version of the above.

**Comment by ZaydH** [comment]
@BenjaminBossan What do you think of adding a simple constructor to the `MyNet` class that forces `criterion__reduce` to `False` similar to below.

```
def __init__(self, *args, **kwargs):
    kwargs["criterion__reduce"] = False
    super().__init__(*args, **kwargs)
```

That way the user never has to remember to set it.

**Comment by BenjaminBossan** [comment]
You could do that, though I would change the code a little bit:

```python
def __init__(self, *args, criterion__reduce=False, **kwargs):
    super().__init__(*args, criterion__reduce=criterion__reduce, **kwargs)
```

**Comment by ZaydH** [comment]
Yours is better.  Would you be fine with a pull request to the docs?  I could submit it.

**Comment by BenjaminBossan** [comment]
Please go ahead.

**Comment by jyhong836** [comment]
You can directly pass `pos_weight` parameter to the criterion, e.g., the `BCEWithLogitsLoss`. Though it weights the positive samples only, it could work in some cases
